File: JavaScript/ExampleTests/common/CatalogItemString.js
import * as tokens from "./tokens.js";

export class CatalogItemString {
  static DNC_STRING = "?DNC?";

  constructor(name = "", price = "") {
    this.name = name;
    this.price = price;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new CatalogItemString(
      v[0] ?? "",
      v[1] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 2, "CatalogItem");
    return new CatalogItemString(
      parts[0],
      parts[1]
    );
  }

  toString() {
    return tokens.token(this.name) + " " + tokens.token(this.price);
  }

  equals(other) {
    if (!(other instanceof CatalogItemString)) return false;
    return (this.name === CatalogItemString.DNC_STRING || other.name === CatalogItemString.DNC_STRING || this.name === other.name)
      && (this.price === CatalogItemString.DNC_STRING || other.price === CatalogItemString.DNC_STRING || this.price === other.price);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/CatalogItemString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class CatalogItemString
    {
        public string name;
        public string price;

        public CatalogItemString(string name, string price)
        {
            this.name = name;
            this.price = price;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static CatalogItemString FromText(string text)
        {
            var parts = Tokens.Require(text, 2, "CatalogItem");
            return new CatalogItemString(parts[0], parts[1]);
        }

        public CatalogItemString(string text)
        {
            var parsed = FromText(text);
            this.name = parsed.name;
            this.price = parsed.price;
        }

        public CatalogItemTyped ToCatalogItemTyped()
        {
            return new CatalogItemTyped(
                new SimpleText(this.name),
                new Dollar(this.price)
            );
        }

        public override string ToString()
        {
            return Tokens.Token(name) + " " + Tokens.Token(price);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not CatalogItemString other) return false;
            return (DNCString == this.name || DNCString == other.name || object.Equals(this.name, other.name))
                && (DNCString == this.price || DNCString == other.price || object.Equals(this.price, other.price));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.name);
            h.Add(this.price);
            return h.ToHashCode();
        }
    }
}
